xarray/core/dataset.py

```diff
@@ -3479,34 +3482,75 @@ def drop(self, labels, dim=None, *, errors="raise"):  # noqa: F811
             any of the variable or index labels passed are not
             in the dataset. If 'ignore', any given labels that are in the
             dataset are dropped and no error is raised.
+        **labels_kwargs : {dim: label, ...}, optional
+            The keyword arguments form of ``dim`` and ``labels``.
 
         Returns
         -------
         dropped : Dataset
+
+        Examples
+        --------
+        >>> data = np.random.randn(2, 3)
+        >>> labels = ['a', 'b', 'c']
+        >>> ds = xr.Dataset({'A': (['x', 'y'], data), 'y': labels})
+        >>> ds.drop(y=['a', 'c'])
+        <xarray.Dataset>
+        Dimensions:  (x: 2, y: 1)
+        Coordinates:
+          * y        (y) <U1 'b'
+        Dimensions without coordinates: x
+        Data variables:
+            A        (x, y) float64 -0.3454 0.1734
+        >>> ds.drop(y='b')
+        <xarray.Dataset>
+        Dimensions:  (x: 2, y: 2)
+        Coordinates:
+          * y        (y) <U1 'a' 'c'
+        Dimensions without coordinates: x
+        Data variables:
+            A        (x, y) float64 -0.3944 -1.418 1.423 -1.041
         """
         if errors not in ["raise", "ignore"]:
             raise ValueError('errors must be either "raise" or "ignore"')
 
-        if dim is None:
+        labels_are_coords = isinstance(labels, DataArrayCoordinates)
+        if labels_kwargs or (utils.is_dict_like(labels) and not labels_are_coords):
```

Comment:
I wonder if it would be more predictable to special case the `dict` type rather than `not labels_are_coords`, which is  just one of many internal xarray mapping types, e.g.,
```suggestion
        if labels_kwargs or isinstance(labels, dict):
```

More generally, I get that writing `array.drop(array.coords)` is convenient, but in the long term it seems like a bad idea to treat different mapping types differently. This suggests that such uses should be deprecated. Either `array.drop_vars(array.coords)` or `array.drop(list(array.coords))` would be less ambiguous.

Reply:
Yeah, I thought needing the check for `labels_are_coords` was odd, but I didn't expect passing in coordinates to take a separate code path. Special-casing the `dict` type makes sense. And the API that is being deprecated is treating `DataArrayCoordinates` as non-dict-like, right?

Reply:
> And the API that is being deprecated is treating `DataArrayCoordinates` as non-dict-like, right?

That's right, we would rather *always* use this  new rule for dict-like inputs.